pumpkin-inventory/src/player.rs

```diff
@@ -84,23 +91,78 @@ impl PlayerInventory {
             _ => unreachable!(),
         }
     }
-
+    pub fn get_slot(&mut self, slot: usize) -> Result<&mut Option<ItemStack>, InventoryError> {
+        match slot {
+            0 => {
+                // TODO: Add crafting check here
+                Ok(&mut self.crafting_output)
+            }
+            1..=4 => Ok(&mut self.crafting[slot - 1]),
+            5..=8 => Ok(&mut self.armor[slot - 5]),
+            9..=44 => Ok(&mut self.items[slot - 9]),
+            45 => Ok(&mut self.offhand),
+            _ => Err(InventoryError::InvalidSlot),
+        }
+    }
     pub fn set_selected(&mut self, slot: usize) {
         assert!((0..9).contains(&slot));
         self.selected = slot;
     }
 
-    pub fn held_item(&self) -> Option<&Item> {
+    pub fn held_item(&self) -> Option<&ItemStack> {
         debug_assert!((0..9).contains(&self.selected));
         self.items[self.selected + 36 - 9].as_ref()
     }
 
-    pub fn slots(&self) -> Vec<Option<&Item>> {
+    pub fn slots(&self) -> Vec<Option<&ItemStack>> {
         let mut slots = vec![self.crafting_output.as_ref()];
         slots.extend(self.crafting.iter().map(|c| c.as_ref()));
         slots.extend(self.armor.iter().map(|c| c.as_ref()));
         slots.extend(self.items.iter().map(|c| c.as_ref()));
         slots.push(self.offhand.as_ref());
         slots
     }
+
+    pub fn slots_mut(&mut self) -> Vec<&mut Option<ItemStack>> {
+        let mut slots = vec![&mut self.crafting_output];
+        slots.extend(self.crafting.iter_mut());
+        slots.extend(self.armor.iter_mut());
+        slots.extend(self.items.iter_mut());
+        slots.push(&mut self.offhand);
+        slots
+    }
+}
+
+impl Container for PlayerInventory {
+    fn window_type(&self) -> &'static WindowType {
+        &WindowType::Generic9x1
+    }
+
+    fn handle_item_change(
+        &mut self,
+        carried_slot: &mut Option<ItemStack>,
+        slot: usize,
+        mouse_click: MouseClick,
+    ) -> Result<(), InventoryError> {
+        let item_slot = self.get_slot(slot)?;
+
+        handle_item_change(carried_slot, item_slot, mouse_click);
+        Ok(())
+    }
+
+    fn all_slots(&mut self) -> Vec<&mut Option<ItemStack>> {
+        self.slots_mut()
+    }
+
+    fn all_slots_ref(&self) -> Vec<Option<&ItemStack>> {
+        self.slots()
+    }
+
+    fn all_combinable_slots(&self) -> Vec<Option<&ItemStack>> {
+        self.items.iter().map(|item| item.as_ref()).collect()
+    }
+
+    fn all_combinable_slots_mut(&mut self) -> Vec<&mut Option<ItemStack>> {
+        self.items.iter_mut().collect()
+    }
```

Comment:
Does this really need to be an Vec?

Reply:
It gets combined with a separate `Vec<&mut Option<ItemStack>>` later, so we'd have to convert it into a Vec anyways